### PR TITLE
Create an analyzer that computes executable lines metrics

### LINKED ISSUE #871: Create an analyzer that computes executable lines metrics
Implements #155

### PR DESCRIPTION
**Pull request by michalb-sonar**
Fix #871

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/Metrics.cs [hunk 1]
```diff
@@ -25,12 +25,15 @@
...
 using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.Common;
 using SonarAnalyzer.Helpers;
 
-namespace SonarAnalyzer.Common.CSharp
+namespace SonarAnalyzer.Metrics.CSharp
 {
     public class Metrics : MetricsBase
     {
+        private ExecutableLinesCalculator executableLinesCalculator = new ExecutableLinesCalculator();
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/Metrics.cs]: MAJOR Make 'executableLinesCalculator' 'readonly'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2933)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Metrics.CSharp
+{
+    public class ExecutableLinesCalculator
+    {
+        public int CountLines(SyntaxTree tree)
+        {
+            var nodesToVisit = new Queue<SyntaxNode>(tree.GetRoot().ChildNodes());
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs]: Not sure to understand why you take only top level nodes and you will take sub level ones after?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            var executableLineNumbers = new HashSet<int>();
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs]: Might sound stupid but 2 statements on the same line will count as only one executable line, is it the expected behavior?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+            while (nodesToVisit.Count > 0)
+            {
+                Visit(nodesToVisit, executableLineNumbers);
+            }
+
+            return executableLineNumbers.Count;
+        }
+
+        private static void Visit(Queue<SyntaxNode> nodesToVisit, HashSet<int> executableLineNumbers)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs]: [ignore] Mutation horror!

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        {
+            var node = nodesToVisit.Dequeue();
+
+            foreach (var n in node.ChildNodes())
+            {
+                nodesToVisit.Enqueue(n);
+            }
+
+            var kind = node.Kind();
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs]: I don't think the extra variable is really needed here.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+                case SyntaxKind.ObjectInitializerExpression:
+                    EnqeueChildExpressions(node, nodesToVisit);
+                    return;
+
+                default:
+                    return;
+            }
+        }
+
+        private static void EnqeueChildExpressions(SyntaxNode node, Queue<SyntaxNode> nodesToVisit)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs]: I would use `InitializerExpressionSyntax` instead of `SyntaxNode` here.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/Metrics.cs [hunk 7]
```diff
@@ -25,12 +25,15 @@
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CSharp;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.Common;
 using SonarAnalyzer.Helpers;
 
-namespace SonarAnalyzer.Common.CSharp
+namespace SonarAnalyzer.Metrics.CSharp
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/Metrics.cs]: Would probably deserve to be in a separate commit.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Metrics/Metrics.cs [hunk 8]
```diff
@@ -80,7 +80,12 @@ protected override bool IsFunction(SyntaxNode node)
 
         public override int GetCognitiveComplexity(SyntaxNode node)
         {
-            return 0;
+            return 0; // Not implemented
+        }
+
+        public override int GetExecutableLinesCount()
+        {
+            return 0; // Not implemented
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Metrics/Metrics.cs]: Could you double check with someone on SonarQube side this is the expected behavior?

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Metrics/ExecutableLinesCalculatorTest.cs [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+using FluentAssertions;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace SonarAnalyzer.UnitTest.Common
+{
+    [TestClass]
+    public class ExecutableLinesCalculatorTest
+    {
+        private static void AssertLinesOfCode(int expectedLineCount, string code)
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Metrics/ExecutableLinesCalculatorTest.cs]: Let's invert parameters order and call it with `expectedLineCount: 0` instead of only `0`.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/SonarAnalyzer.UnitTest.csproj [hunk 10]
```diff
@@ -156,6 +156,7 @@
     </None>
     <Compile Include="Helpers\FlowAnalysis\*.cs" />
     <Compile Include="Helpers\*.cs" />
+    <Compile Include="Metrics\ExecutableLinesCalculatorTest.cs" />
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/SonarAnalyzer.UnitTest.csproj]: Doesn't match the pattern used by other lines.

File: sonaranalyzer-dotnet/its/regression-test.ps1 [hunk 11]
```diff
@@ -6,7 +6,7 @@ param
     $msbuildVersion = "14.0",
 
     [Parameter(HelpMessage = "The key of the rule to test, e.g. S1234. If omitted, all rule will be tested.")]
-    [ValidatePattern("^S[0-9]+$")]
+    [ValidatePattern("^S[0-9]+")]
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/its/regression-test.ps1]: Should be in a different commit.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Metrics.CSharp
+{
+    public class ExecutableLinesCalculator
```
[the remark this answers is not in the document]
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs]: Tried that. It was much less readable.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs (hunk 2 above)
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs]: Oh yes, I can use DescendantNodes instead.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs (hunk 3 above)
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs]: Yes. This is what VS metrics do and in the spirit of what Ann's article suggests.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs (hunk 5 above)
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs]: I decided to leave it, as otherwise debugging is painful - you have to click in the variable inspector, so the method is re-evaluated.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs]: Except that you have more methods I don't think it will be worse but that's my view.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Metrics/ExecutableLinesCalculatorTest.cs (hunk 9 above)
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Metrics/ExecutableLinesCalculatorTest.cs]: It is pretty clear from the name and both arguments what the method does. Also, having expected as a first parameter is a common convention.

It's not fluent, but to me that looks better than your suggestion.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs]: For example this
```
case SyntaxKind.IfStatement:
    CommonPart();
```
Would look like that:
```
public override void VisitIfStatement(IfStatementSyntax node)
{
     CommonPart();
     base.VisitIfStatement(node);
}
```
Multiply it by number of cases and the class becomes much bigger - especially given the action we take is always the same.

[the remark this answers is not in the document]
Comment (duncanp-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs]: I don't think you need to do that - you could just override _DefaultVisit_.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs (hunk 2 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs]: At some point we will need to make some performance test. @dbolkensteyn told me it was recommended by roslyn team not to use methods querying children and then using Linq but to rather subscribe to the different levels we are interested in.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Metrics/ExecutableLinesCalculatorTest.cs [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+        public void Test_06_Loops()
+        {
+            AssertLinesOfCode(4,
+              @"
+                void Foo(int[] arr)
+                {
+                    do {} while (true); // +1
+                    foreach (var a in arr) { }// +1
+                    for (;;) { } // +1
+                    while (true) { } // +1
```
[the remark this answers is not in the document]
Comment (valhristov) [reply sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Metrics/ExecutableLinesCalculatorTest.cs]: imo, this
```
while
(true)
{
}
```
should be two or three, depending if you count the opening brace.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs (hunk 2 above)
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs]: Interesting!

I've run the test (with walker and overriding `DefaultVisit`). The walker was quicker by 30ms, with total time of about 0.5s during ITs. That's probably within a margin of error.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Metrics/ExecutableLinesCalculatorTest.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Metrics/ExecutableLinesCalculatorTest.cs]: I'm loosely following VS metrics, which show this as 1.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs (hunk 3 above)
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesCalculator.cs]: @duncanp-sonar can you provide a short code sample?

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Metrics/ExecutableLinesCalculatorTest.cs [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+
+
+        [TestMethod]
+        public void Test_06_Loops()
+        {
+            AssertLinesOfCode(4,
+              @"
+                void Foo(int[] arr)
+                {
+                    do {} while (true); // +1
```
[the remark this answers is not in the document]
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/Metrics/ExecutableLinesCalculatorTest.cs]: That would count 1. I've added a test for that.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesWalker.cs [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Metrics.CSharp
+{
+    public class ExecutableLinesWalker : CSharpSyntaxWalker
+    {
+        private HashSet<int> executableLineNumbers = new HashSet<int>();
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/ExecutableLinesWalker.cs]: MAJOR Make 'executableLineNumbers' 'readonly'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2933)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Metrics/Metrics.cs (hunk 8 above)
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Metrics/Metrics.cs]: Didn't get the response back yet, will chase it up on Monday.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/Metrics.cs (hunk 7 above)
Comment (michalb-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Metrics/Metrics.cs]: Not worth spliting now, but I will keep it in mind for future.
